src/coreclr/scripts/genUserEvents.py

```diff
@@ -0,0 +1,407 @@
+            pack_list.append("    TraceLoggingBinary(%s, %s)" % (parameter, size))
+        elif paramName in template.arrays:
+            size = "sizeof(%s) * (int)%s" % (
+                getArrayDataTypeMapping(runtimeFlavor)[parameter.winType],
+                parameter.prop)
+        elif parameter.winType == "win:GUID":
+            pack_list.append("    %s((uint8_t *)%s)" % (getUserEventDataTypeMapping(runtimeFlavor)[parameter.winType], parameter.name))
+        else:
+            pack_list.append("    %s(%s)" % (getUserEventDataTypeMapping(runtimeFlavor)[parameter.winType], parameter.name))
+    code = ",\n".join(pack_list)  + "\n    );\n    return ERROR_SUCCESS;\n" #+ "\n\n"
+    return code
+
+
+
+keywordMap = {}
+
+def generateEventKeywords(eventKeywords):
+    mask = 0
+    # split keywords if there are multiple
+    allKeywords = eventKeywords.split()
+
+    for singleKeyword in allKeywords:
+        mask = mask | keywordMap[singleKeyword]
+
+    return mask
+
+def getCoreCLRUserEventHelperFileImplPrefix():
+    return """
+#include "common.h"
+#include <stdlib.h>
+#include <string.h>
+
+#ifndef TARGET_UNIX
+#include <windef.h>
+#include <crtdbg.h>
+#else
+#include "pal.h"
+#endif //TARGET_UNIX
+
+"""
+
+def generateUserEventHelperFile(etwmanifest, userevent_directory, target_cpp, runtimeFlavor, extern, dryRun):
+    usereventhelpersPath = os.path.join(userevent_directory, "usereventshelpers" + (".cpp" if target_cpp else ".c"))
+    if dryRun:
+        print(usereventhelpersPath)
+    else:
+        with open_for_update(usereventhelpersPath) as helper:
+            helper.write(stdprolog_cpp)
+            helper.write("""
+#include <user_events.h>
+#include <common.h>
+#include <configuration.h>
+
+bool s_userEventsEnabled = false;
+
+void InitUserEvents()
+{
+    // TODO: What to call this runtimeconfig.json switch? S.D.T implies it supports EventSource
+    bool isEnabled = Configuration::GetKnobBooleanValue(W("System.Diagnostics.Tracing.UserEvents"), false); 
+    if (!isEnabled)
+    {
+        isEnabled = CLRConfig::GetConfigValue(CLRConfig::INTERNAL_EnableUserEvents) != 0;
+    }
+
+    s_userEventsEnabled = isEnabled;
+}
+
+bool IsUserEventsEnabled()
+{
+    return s_userEventsEnabled;
+}
+""")
+
+def getCoreCLRUserEventImplFilePrefix():
+    return """#include "common.h"
+#include <stdint.h>
+#include <eventheader/TraceLoggingProvider.h>
+#include <user_events.h>
+"""
+
+def getCoreCLRUserEventImplFileSuffix():
+    return ""
+
+def getMonoUserEventImplFilePrefix():
+    return ""
+
+def getMonoUserEventImplFileSuffix():
+    return ""
+
+def getAotUserEventImplFilePrefix():
+    return ""
+
+def getAotUserEventImplFileSuffix():
+    return ""
+
+def generateUserEventImplFiles(
+        etwmanifest, userevent_directory, extern, target_cpp, runtimeFlavor, inclusionList, exclusionList, dryRun):
+    tree = DOM.parse(etwmanifest)
+
+    # Find the src directory starting with the assumption that
```

Comment:
I don't understand why the script is searching for the src directory? Maybe I'm missing something but it looks like after it locates the src directory it never uses it?

Reply:
This is an artifact because we started with genEventPipe.py and modified to work with user_events. I don't think it has a purpose in either file